agents/packages/ethlisbon/skills/price_prophet/behaviours.py

```diff
@@ -226,25 +214,53 @@ def async_act(self) -> Generator:
 class TrainModelBehaviour(PriceProphetBaseBehaviour):
     """TrainModelBehaviour"""
 
-    # TODO: set the following class attributes
-    state_id: str
     behaviour_id: str = "train_model"
     matching_round: Type[AbstractRound] = TrainModelRound
 
-    # TODO: implement logic required to set payload content (e.g. synchronized_data)
     def async_act(self) -> Generator:
         """Do the act, supporting asynchronous execution."""
 
+        # TODO: get randomness from randomness beacon
         with self.context.benchmark_tool.measure(self.behaviour_id).local():
-            sender = self.context.agent_address
-            payload = TrainModelPayload(sender=sender, content=...)
+            random_state = self.synchronized_data.period_count
+            most_voted: JSONLike = self.get_strict(AnnotateDataRound.selection_key)
+            results_grid = self.train_model(pd.read_json(most_voted), random_state)
+            content, sender = results_grid.to_json(), self.context.agent_address
+            payload = TrainModelPayload(sender=sender, content=content)
 
         with self.context.benchmark_tool.measure(self.behaviour_id).consensus():
             yield from self.send_a2a_transaction(payload)
             yield from self.wait_until_round_end()
 
         self.set_done()
 
+    def train_model(self, x, random_state):
+        """Train model"""
+
+        # currently we interpolate and discard features with NaN
+        x_imputed = impute(x)
+        x_train, x_test = split_train_test(x_imputed)
+        forecaster.regressor.random_state = random_state
+
+        # this runs the grid search
+        # currently still univariate (without exogenous variables)
+        results_grid = grid_search_forecaster(
+            forecaster=forecaster,
+            y=x_train[Y_TARGET],
+            param_grid=PARAM_GRID,
+            lags_grid=LAGS_GRID,
+            steps=STEPS_INTO_THE_FUTURE,
+            refit=True,  # set False to speed up
+            metric="mean_squared_error",
+            initial_train_size=x_train.shape[0] // 2,
+            fixed_train_size=False,
+            return_best=True,
+            verbose=False,
+        )
```

Comment:
can we set some seed here? this might cause consensus failures, as i think this would produce different results in different agents. Alternatively we could change the round type to `CollectDifferentUntilThreshold`

Reply:
https://github.com/8ball030/eth_lisbon/blob/96724ca3289d03304b8fd5872fec395af7931020/agents/packages/ethlisbon/skills/price_prophet/behaviours.py#L223-L225

https://github.com/8ball030/eth_lisbon/blob/96724ca3289d03304b8fd5872fec395af7931020/agents/packages/ethlisbon/skills/price_prophet/behaviours.py#L243